# ğŸ§¬ RNA-seq_Data_Visualization_R-Shiny-App 

An interactive R Shiny application for exploring and visualizing RNA-seq data.  
This app was developed as part of coursework and lab projects at Boston University to provide a user-friendly interface for analyzing RNA-seq datasets.

---

## âš™ï¸ Features

The app is organized into the following four main analysis modules:

1. **Metadata Analysis**  
   - Upload and summarize sample metadata.
   - This data can be uploaded in .csv format.  
   - Explore data tables and generate density plots by selected columns and groups.  

2. **Counts Matrix Analysis**  
   - Upload a normalized counts matrix in .csv format.  
   - Filter genes based on variance percentile and non-zero sample thresholds.  
   - View diagnostic plots, clustered heatmaps, and PCA visualizations (scatterplots or beeswarm).  

3. **Differential Expression Analysis**  
   - Upload DESeq2 results in .csv format.  
   - Interactively adjust thresholds for log2 fold change and adjusted p-value.  
   - Visualize results via p-value histograms, log2FC histograms, volcano plots, and top-gene expression plots.  

4. **FGSEA (Gene Set Enrichment) Analysis**  
   - Upload FGSEA results (CSV/TSV).  
   - Explore top enriched pathways with customizable filters.  
   - Visualize results with barplots and scatterplots.  
   - Download filtered tables for further use.  

---

## ğŸ’»ğŸ“¦ Installation & Requirements

The app requires **R (version â‰¥ 4.0)** and the following packages:

- `shiny`  
- `DT`  
- `ggplot2`  
- `dplyr`  
- `pheatmap`  
- `reshape2`  
- `stats`  
- `ggbeeswarm`  
- `gridExtra`  
- `viridis`

These can be installed in R using:

install.packages(c("shiny", "DT", "ggplot2", "dplyr", 
                   "pheatmap", "reshape2", "ggbeeswarm", 
                   "gridExtra", "viridis"))

---


## â–¶ï¸ Usage

1. First clone this reporsitory by running the following in your terminal:
    git clone https://github.com/Saneeya1402/RNAseq-ShinyApp.git
    cd RNAseq-ShinyApp
2. Then open R and run the app like:
   library(shiny)
   runApp("app.R")
3. Using the app's inerface, upload the following documents in their respective tabs:
- Metadata file (.csv)
- Normalized counts matrix (.csv)
- DESeq2 results (.csv)
- FGSEA results (.csv, .tsv, or .txt)


---


## ğŸ“ Repository Structure

RNAseq-ShinyApp/
â”‚
â”œâ”€â”€ app.R              # Main Shiny application
â”œâ”€â”€ data/              # (Optional) Example input data
â”œâ”€â”€ www/               # (Optional) Assets (CSS/JS/images)
â””â”€â”€ README.md          # Project documentation


---


## ğŸ“ŠğŸ–¼ï¸ Example Outputs:

- Density plots of metadata columns
- Gene filtering diagnostics
- Clustered heatmaps of filtered genes
- PCA scatterplots / beeswarm plots
- Volcano plots and DEG summaries
- FGSEA top pathway barplots and scatterplots

